source/common/http/http1/codec_impl.cc

```diff
@@ -463,22 +429,22 @@ void ConnectionImpl::onHeaderField(const char* data, size_t length) {
   current_header_field_.append(data, length);
 }
 
-void ConnectionImpl::onHeaderValue(const char* data, size_t length) {
+void ConnectionImpl::onHeaderValueBase(const char* data, size_t length) {
   if (header_parsing_state_ == HeaderParsingState::Done) {
     // Ignore trailers.
     return;
   }
 
   const absl::string_view header_value = absl::string_view(data, length);
-
   if (strict_header_validation_) {
     if (!Http::HeaderUtility::headerIsValid(header_value)) {
       ENVOY_CONN_LOG(debug, "invalid header value: {}", connection_, header_value);
       error_code_ = Http::Code::BadRequest;
       sendProtocolError();
       throw CodecProtocolException("http/1.1 protocol error: header value contains invalid chars");
     }
-  } else if (header_value.find('\0') != absl::string_view::npos) {
+  } else if (ParserFactory::usesLegacyParser() &&
```

Comment:
Is this an optimization because we're confident the new parser handles this for us?  I'd think we wouldn't want to allow \0 for the new version either.
(calling out because if we make this a runtime switch we can't depend on the previously stable value)